libs/sdk-utils/multicall/src/lib/sdk-utils-multicall.ts

```diff
@@ -1,23 +1,70 @@
-export function sdkUtilsMulticall(): string {
-  return 'sdk-utils-multicall';
-}
-
+import { ContractWrapper, decorate } from '@sdk-utils/core';
+import { providers as _0xS } from '@0xsequence/multicall';
+import { Contract, ethers } from 'ethers';
+import { Provider } from '@ethersproject/providers';
+import { MulticallProvider } from '@0xsequence/multicall/dist/declarations/src/providers';
 
 /**
- * MyType extends a `thing` that does some `stuff`
+ * Multicall implementation with 0xSequence happens automatically.
+ * There are no special methods to call, but we add this flag purely as an indicator to show the contract has been wrapped.
  */
-interface MyType {
-  // the value of choice
-  myValue: string
-}
+type Multicall = {
+  isMulticallEnabled: boolean;
+};
 
+/**
+ * We need to extend the internal provider to get type inference on multicall provider properties.
+ */
+type WrapMulticallProvder<C extends Contract> = C & {
+  provider: MulticallProvider;
+};
 
 /**
- * do stuff is a great function
+ * A list of contract addresses for multicall, on different networks.
+ *
+ * Visit [The 0xSequence Multicall page](https://github.com/0xsequence/sequence.js/tree/master/packages/multicall) for details.
  */
-export async function doStuff(input: unknown): Promise<MyType> {
+export const MULTICALLCONTRACTS = {
+  // Put deployed contracts here
+  1: '0xd130B43062D875a4B7aF3f8fc036Bc6e9D3E1B3E',
+};
+
+export class MultiCallWrapper extends ContractWrapper<Multicall> {
+  // Set in the constructor through the set function
+  private multicallProvider!: _0xS.MulticallProvider;
+  public contracts: Contract[] = [];
+
+  constructor(provider?: Provider) {
+    super();
+    this.setMulticallProvider(provider);
+  }
+
+  /**
+   * Wraps the existing provider in the 0x Multicall provider.
+   * Batch calls together with Promise.all to execute through multicall.
+   * Alternatively, see the [The 0xSequence Multicall page](https://github.com/0xsequence/sequence.js/tree/master/packages/multicall) page for usage details.
+   * @param provider an ethers provider.
+   */
+  private setMulticallProvider(provider: undefined | Provider) {
```

Comment:
Why do we allow `undefined` to be passed here?

Reply:
Because the base behavior is to wrap the ethers.defaultProvider as a fallback. Hence, the provider is not strictly necessary.